crates/stateless/src/validation.rs

```diff
@@ -0,0 +1,292 @@
+
+    /// Error when required ancestor headers are missing (e.g., parent header for pre-state root).
+    #[error("missing required ancestor headers")]
+    MissingAncestorHeader,
+
+    /// Error when deserializing ancestor headers
+    #[error("could not deserialize ancestor headers")]
+    HeaderDeserializationFailed,
+
+    /// Error when the computed state root does not match the one in the block header.
+    #[error("mismatched post- state root: {got}\n {expected}")]
+    PostStateRootMismatch {
+        /// The computed post-state root
+        got: B256,
+        /// The expected post-state root; in the block header
+        expected: B256,
+    },
+
+    /// Error when the computed pre-state root does not match the expected one.
+    #[error("mismatched pre-state root: {got} \n {expected}")]
+    PreStateRootMismatch {
+        /// The computed pre-state root
+        got: B256,
+        /// The expected pre-state root from the previous block
+        expected: B256,
+    },
+}
+
+/// Performs stateless validation of a block using the provided witness data.
+///
+/// This function attempts to fully validate a given `current_block` statelessly, ie without access
+/// to a persistent database.
+/// It relies entirely on the `witness` data and `ancestor_headers`
+/// provided alongside the block.
+///
+/// The witness data is validated in the following way:
+///
+/// 1. **Ancestor Header Verification:** Checks if the `ancestor_headers` are present, form a
+///    contiguous chain back from `current_block`'s parent, and do not exceed the `BLOCKHASH` opcode
+///    limit using `compute_ancestor_hashes`. We must have at least one ancestor header, even if the
+///    `BLOCKHASH` opcode is not used because we need the state root of the previous block to verify
+///    the pre state reads.
+///
+/// 2. **Pre-State Verification:** Retrieves the expected `pre_state_root` from the parent header
+///    from `ancestor_headers`. Verifies the provided [`ExecutionWitness`] against this root using
+///    [`verify_execution_witness`].
+///
+/// 3. **Chain Verification:** The code currently does not verify the [`ChainSpec`] and expects a
+///    higher level function to assert that this is correct by, for example, asserting that it is
+///    equal to the Ethereum Mainnet `ChainSpec` or asserting against the genesis hash that this
+///    `ChainSpec` defines.
+///
+/// High Level Overview of functionality:
+///
+/// - Verify all state accesses against a trusted pre-state root
+/// - Put all state accesses into an in-memory database
+/// - Use the in-memory database to execute the block
+/// - Validate the output of block execution (e.g. receipts, logs, requests)
+/// - Compute the post-state root using the state-diff from block execution
+/// - Check that the post-state root is the state root in the block.
+///
+/// If all steps succeed the function returns `Some` containing the hash of the validated
+/// `current_block`.
+pub fn stateless_validation(
```

Comment:
i don't see any equivalent of pre-execution validation. for ref see https://github.com/paradigmxyz/reth/blob/f0d3d1d32ae8ac0b1005f20ad4291a0661eac66d/crates/engine/tree/src/tree/mod.rs#L1912-L1938

Reply:
Done! Good catch :D